nca/Parsers/IstioVirtualServiceYamlParser.py

```diff
@@ -0,0 +1,361 @@
+#
+# Copyright 2020- IBM Inc. All rights reserved
+# SPDX-License-Identifier: Apache2.0
+#
+
+import re
+from functools import reduce
+from nca.CoreDS.MinDFA import MinDFA
+from nca.CoreDS.MethodSet import MethodSet
+from nca.Resources.OtherResources.VirtualService import VirtualService
+from .GenericIngressLikeYamlParser import GenericIngressLikeYamlParser
+
+
+class IstioVirtualServiceYamlParser(GenericIngressLikeYamlParser):
+    """
+    A parser for Istio VirtualService resources.
+    """
+
+    def __init__(self, peer_container):
+        """
+        :param PeerContainer peer_container: The ingress policy will be evaluated
+        against this set of peers
+        """
+        GenericIngressLikeYamlParser.__init__(self, peer_container)
+        self.virtual_services = {}  # a map from a name to a VirtualService
+        self.gtw_parser = None
+
+    def add_virtual_service(self, vs):
+        """
+        Adds a virtual service by its full name
+        :param VirtualService vs: the virtual service to add
+        """
+        self.virtual_services[vs.full_name()] = vs
+
+    def parse_virtual_service(self, vs_resource, vs_file_name):
+        """
+        Parses a virtual service resource object and adds the parsed VirtualService object to self.virtual_services
+        :param dict vs_resource: the virtual service object to parse
+        :param str vs_file_name: the name of the virtual service resource file
+        """
+        self.set_file_name(vs_file_name)  # for error/warning messages
+        vs_name, vs_ns = self.parse_generic_yaml_objects_fields(vs_resource, ['VirtualService'],
+                                                                ['networking.istio.io/v1alpha3',
+                                                                 'networking.istio.io/v1beta1'], 'istio', True)
+        if vs_name is None:
+            return None  # Not an Istio VirtualService object
+        vs_namespace = self.peer_container.get_namespace(vs_ns)
+        vs = VirtualService(vs_name, vs_namespace)
+
+        vs_spec = vs_resource['spec']
+        self.check_fields_validity(vs_spec, f'VirtualService {vs.full_name()}',
+                                   {'hosts': [0, list], 'gateways': [0, list], 'http': 0, 'tls': 0, 'tcp': 0,
+                                    'exportTo': [3, list]})
+        hosts = vs_spec.get('hosts')
```

Comment:
```suggestion
                # field 'hosts' is the destination hosts to which traffic is being sent (from the original http request)
                # (see https://github.com/istio/api/blob/bb3cb9c034df2b5cc1de1d77689d201a0cf961c5/networking/v1alpha3/virtual_service.proto#L209-L238 )
                hosts = vs_spec.get('hosts')
```

Reply:
do we have any assumptions on the hosts? (on egress gateway analysis it has to match a serviceEntry dns address?)

Reply:
We only use the `hosts` field for matching the virtual service with gateways. We also add the matched hosts to the `hosts` attribute of the connections. 
For destination (either local or dns addresses), we use the `destination.host` field.